api/src/infrastructure/postgres-survey-data.repository.ts

```diff
@@ -0,0 +1,107 @@
+import { DataSource, EntityManager } from 'typeorm';
+import { ISurveyDataRepository } from '@api/infrastructure/survey-data-repository.interface';
+import { Inject, Logger } from '@nestjs/common';
+import { InjectDataSource } from '@nestjs/typeorm';
+import { WidgetDataFilters } from '@shared/dto/widgets/widget-data-filter';
+import { SectionWithDataWidget } from '@shared/dto/sections/section.entity';
+import { SQLAdapter } from '@api/infrastructure/sql-adapter';
+import { WidgetQuestionMap } from '@shared/dto/widgets/widget-question-map';
+import {
+  BaseWidgetWithData,
+  WidgetData,
+} from '@shared/dto/widgets/base-widget-data.interface';
+
+export class PostgresSurveyDataRepository implements ISurveyDataRepository {
+  public constructor(
+    private readonly logger: Logger,
+    @Inject(SQLAdapter) private readonly sqlAdapter: SQLAdapter,
+    @InjectDataSource() private readonly dataSource: DataSource,
+  ) {}
+
+  public async addSurveyDataToSections(
+    sections: SectionWithDataWidget[],
+    filters?: WidgetDataFilters,
+  ): Promise<SectionWithDataWidget[]> {
+    // Perform a transaction to ensure isolation
+    return await this.dataSource.transaction('READ COMMITTED', async (t) => {
+      let answersTable: string = 'survey.answers';
+
+      if (filters !== undefined) {
+        answersTable = `filtered_answers`;
+
+        const filterClause =
+          this.sqlAdapter.generateSqlFromWidgetDataFilters(filters);
+
+        await t.query(`CREATE TEMPORARY TABLE ${answersTable} ON COMMIT DROP AS
+SELECT * FROM survey.answers WHERE survey_id IN (SELECT survey_id FROM survey.answers ${filterClause})`);
+      }
+
+      const widgetDataPromises = [];
+      for (let sectionIdx = 0; sectionIdx < sections.length; sectionIdx++) {
+        const section = sections[sectionIdx];
+        const baseWidgets = section.baseWidgets;
+        for (let widgetIdx = 0; widgetIdx < baseWidgets.length; widgetIdx++) {
+          const widget = baseWidgets[widgetIdx];
+          widgetDataPromises.push(
+            this.appendBaseWidgetData(t, answersTable, widget),
+          );
+        }
+      }
+
+      await Promise.all(widgetDataPromises);
+      return sections;
+    });
+  }
+
+  private async appendBaseWidgetData(
+    t: EntityManager,
+    answersTable: string,
+    widget: BaseWidgetWithData,
+  ): Promise<void> {
+    const question = WidgetQuestionMap.getQuestionByWidgetIndicator(
+      widget.indicator,
+    );
+    if (question === undefined) {
+      this.logger.warn(
+        `${widget.indicator} not found in WidgetQuestionMap. Skipping filter...`,
+        this.constructor.name,
+      );
+      return;
+    }
+    // TODO: Edge cases here. Total surveys and total countries.
+    if (widget.indicator === 'Total surveys') {
```

Comment:
Below code is again evil AF. We need to avoid raw queries using magic strings every time it's possible to use a queryBuilder.

The point of using a querybuilder is that we get type safety to perform queries, when pointing to specific properties of specific tables. Maintaining this code is hard, we have to avoid magic strings